libsolidity/analysis/OverrideChecker.cpp

```diff
@@ -474,56 +674,49 @@ void OverrideChecker::checkAmbiguousOverridesInternal(set<
 	} cutVertexFinder{overrideGraph};
 
 	// Remove all base functions overridden by cut vertices (they don't need to be overridden).
-	for (auto const* function: cutVertexFinder.cutVertices())
+	for (OverrideProxy const& function: cutVertexFinder.cutVertices())
 	{
-		std::set<CallableDeclaration const*> toTraverse = function->annotation().baseFunctions;
+		std::set<OverrideProxy> toTraverse = function.baseFunctions();
 		while (!toTraverse.empty())
 		{
-			auto const *base = *toTraverse.begin();
+			OverrideProxy base = *toTraverse.begin();
 			toTraverse.erase(toTraverse.begin());
 			_baseCallables.erase(base);
-			for (CallableDeclaration const* f: base->annotation().baseFunctions)
+			for (OverrideProxy const& f: base.baseFunctions())
 				toTraverse.insert(f);
 		}
 		// Remove unimplemented base functions at the cut vertices itself as well.
-		if (auto opt = dynamic_cast<ImplementationOptional const*>(function))
-			if (!opt->isImplemented())
-				_baseCallables.erase(function);
+		if (function.unimplemented())
+			_baseCallables.erase(function);
 	}
 
 	// If more than one function is left, they have to be overridden.
 	if (_baseCallables.size() <= 1)
 		return;
 
 	SecondarySourceLocation ssl;
-	for (auto const* baseFunction: _baseCallables)
-	{
-		string contractName = dynamic_cast<ContractDefinition const&>(*baseFunction->scope()).name();
-		ssl.append("Definition in \"" + contractName + "\": ", baseFunction->location());
-	}
+	for (OverrideProxy const& baseFunction: _baseCallables)
+		ssl.append("Definition in \"" + baseFunction.contractName() + "\": ", baseFunction.location());
 
-	string callableName;
-	string distinguishigProperty;
-	if (dynamic_cast<FunctionDefinition const*>(*_baseCallables.begin()))
-	{
-		callableName = "function";
-		distinguishigProperty = "name and parameter types";
-	}
-	else if (dynamic_cast<ModifierDefinition const*>(*_baseCallables.begin()))
-	{
-		callableName = "modifier";
-		distinguishigProperty = "name";
-	}
-	else
-		solAssert(false, "Invalid type for ambiguous override.");
+	string callableName = _baseCallables.begin()->isModifier() ? _baseCallables.begin()->astNodeName() : "function";
```

Comment:
Shouldn't `astNodeName()` work for `function` as well?

Reply:
It should say "function" for both functions and public state variables. Maybe I can make that more explicit.